File: Compiler/Core/CorePrep/Wire/Decode.cpp
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0

#include <limits>
#include <string_view>
#include <type_traits>

#include "Visual/XSharp/Core/CorePrep/Wire.hpp"
#include "Visual/XSharp/Core/Scalar.hpp"

namespace visual_xsharp::core::wire
{
    namespace
    {
        class Reader final
        {
        public:
            Reader(std::span<const std::uint8_t> bytes, const Limits &limits)
                : bytes_(bytes)
                , limits_(limits)
            {}

            [[nodiscard]] auto
            document() -> DecodeResult
            {
                if (bytes_.size() > limits_.maximum_wire_bytes)
                    return failure(ErrorKind::LimitExceeded, "wire byte length", "input exceeds configured byte limit");
                for (const auto expected : magic)
                    if (byte("magic") != expected)
                        return failure(ErrorKind::InvalidMagic, "magic", "input is not a Visual X# CorePrep wire document");
                if (error_)
                    return result();
                const auto version = unsigned_integer<std::uint16_t>("version");
                if (!error_ && version != current_version)
                    fail(ErrorKind::UnsupportedVersion, "version", "unsupported CorePrep wire version");
                const auto flags = unsigned_integer<std::uint16_t>("flags");
                if (!error_ && flags != 0)
                    fail(ErrorKind::InvalidTag, "flags", "reserved wire flags must be zero");
                if (error_)
                    return result();
                CorePrepModule module;
                module.name = qualified_name("module name");
                module.functions = vector<Function>(limits_.maximum_functions, "function count", [this] {
                    return function();
                });
                if (!error_ && offset_ != bytes_.size())
                    fail(ErrorKind::TrailingInput, "document", "bytes remain after CorePrep module");
                if (!error_)
                    module_ = std::move(module);
                return result();
            }

        private:
            std::span<const std::uint8_t> bytes_;
            const Limits &limits_;
            std::size_t offset_{};
            std::optional<CorePrepModule> module_;
            std::optional<Error> error_;

            [[nodiscard]] auto
            result() -> DecodeResult
            {
                return DecodeResult{ std::move(module_), std::move(error_) };
            }
            [[nodiscard]] auto
            failure(ErrorKind kind, std::string context, std::string message) -> DecodeResult
            {
                fail(kind, std::move(context), std::move(message));
                return result();
            }
            void
            fail(ErrorKind kind, std::string context, std::string message)
            {
                if (!error_)
                    error_ = Error{ kind, offset_, std::move(context), std::move(message) };
            }

            [[nodiscard]] auto
            byte(std::string_view context) -> std::uint8_t
            {
                if (offset_ >= bytes_.size())
                {
                    fail(ErrorKind::TruncatedInput, std::string(context), "input ended before the requested byte");
                    return 0;
                }
                return bytes_[offset_++];
            }

            template<typename Integer>
            [[nodiscard]] auto
            unsigned_integer(std::string_view context) -> Integer
            {
                static_assert(std::is_unsigned_v<Integer>);
                Integer result{};
                for (std::size_t shift = 0; shift < sizeof(Integer) * 8U; shift += 8U)
                    result |= static_cast<Integer>(byte(context)) << shift;
                return result;
            }

            [[nodiscard]] auto
            count(std::size_t maximum, std::string_view context) -> std::size_t
            {
                const auto value = unsigned_integer<std::uint32_t>(context);
                if (!error_ && value > maximum)
                    fail(ErrorKind::LimitExceeded, std::string(context), "collection count exceeds configured limit");
                return error_ ? 0U : static_cast<std::size_t>(value);
            }

            template<typename Value, typename Decode>
            [[nodiscard]] auto
            vector(std::size_t maximum, std::string_view context, Decode decode) -> std::vector<Value>
            {
                const auto size = count(maximum, context);
                std::vector<Value> values;
                if (error_)
                    return values;
                values.reserve(size);
                for (std::size_t index = 0; index < size && !error_; ++index)
                    values.push_back(decode());
                return values;
            }

            [[nodiscard]] auto
            text(std::string_view context) -> std::u32string
            {
                const auto size = count(limits_.maximum_string_code_points, context);
                std::u32string value;
                value.reserve(size);
                for (std::size_t index = 0; index < size && !error_; ++index)
                {
                    const auto code_point = unsigned_integer<std::uint32_t>(context);
                    if (code_point > 0x10ffffU || (code_point >= 0xd800U && code_point <= 0xdfffU))
                    {
                        fail(ErrorKind::InvalidCodePoint, std::string(context), "wire text contains a non-scalar Unicode code point");
                        break;
                    }
                    value.push_back(static_cast<char32_t>(code_point));
                }
                return value;
            }

            [[nodiscard]] auto
            qualified_name(std::string_view context) -> std::vector<std::u32string>
            {
                return vector<std::u32string>(65535U, context, [this, context] {
                    return text(context);
                });
            }

            [[nodiscard]] auto
            symbol(std::string_view context) -> SymbolName
            {
                const auto id = unsigned_integer<SymbolId>(context);
                if (!error_ && id == 0)
                    fail(ErrorKind::InvalidSymbol, std::string(context), "symbol id must be positive");
                return SymbolName{ id, text(context) };
            }

            [[nodiscard]] auto
            type(std::size_t depth = 0) -> Type
            {
                if (depth > limits_.maximum_type_depth)
                {
                    fail(ErrorKind::LimitExceeded, "type", "type nesting exceeds configured limit");
                    return Type::unit();
                }
                const auto tag = byte("type tag");
                switch (tag)
                {
                    case 0:
                        return Type::unit();
                    case 1:
                        return Type::boolean();
                    case 2:
                        return Type::int64();
                    case 3:
                        return Type::int32();
                    case 4:
                        return Type::string();
                    case 5:
                    {
                        auto parameters = vector<Type>(65535U, "function type parameter count", [this, depth] {
                            return type(depth + 1U);
                        });
                        auto result = type(depth + 1U);
                        return Type::function(std::move(parameters), std::move(result));
                    }
                    case 6:
                    {
                        auto name = qualified_name("named type");
                        auto arguments = vector<Type>(65535U, "type argument count", [this, depth] {
                            return type(depth + 1U);
                        });
                        return Type::named(std::move(name), std::move(arguments));
                    }
                    case 7:
                        return Type::type_variable(symbol("type variable symbol"));
                    case 8:
                        return Type::character();
                    case 9:
                        return Type::int8();
                    case 10:
                        return Type::int16();
                    case 11:
                        return Type::int128();
                    case 12:
                        return Type::uint8();
                    case 13:
                        return Type::uint16();
                    case 14:
                        return Type::uint32();
                    case 15:
                        return Type::uint64();
                    case 16:
                        return Type::uint128();
                    case 17:
                        return Type::float16();
                    case 18:
                        return Type::float32();
                    case 19:
                        return Type::float64();
                    case 20:
                        return Type::float128();
                    default:
                        fail(ErrorKind::InvalidTag, "type tag", "unknown CorePrep type tag");
                        return Type::unit();
                }
            }

            [[nodiscard]] auto
            boolean(std::string_view context) -> bool
            {
                const auto value = byte(context);
                if (value > 1U)
                    fail(ErrorKind::InvalidBoolean, std::string(context), "boolean byte must be zero or one");
                return value == 1U;
            }

            [[nodiscard]] auto
            literal(const Type &value_type) -> Literal
            {
                switch (value_type.kind)
                {
                    case Type::Kind::Unit:
                        return std::monostate{};
                    case Type::Kind::Bool:
                        return boolean("boolean literal");
                    case Type::Kind::Character:
                    case Type::Kind::Int8:
                    case Type::Kind::Int16:
                    case Type::Kind::Int32:
                    case Type::Kind::Int64:
                    case Type::Kind::Int128:
                    case Type::Kind::UInt8:
                    case Type::Kind::UInt16:
                    case Type::Kind::UInt32:
                    case Type::Kind::UInt64:
                    case Type::Kind::UInt128:
                    {
                        IntegerLiteral value;
                        value.negative = boolean("integer sign");
                        value.magnitude = vector<std::uint8_t>(limits_.maximum_numeric_bytes, "integer magnitude", [this] {
                            return byte("integer magnitude");
                        });
                        if (!integer_is_canonical(value))
                            fail(ErrorKind::InvalidInteger, "integer literal", "integer magnitude/sign is not canonical");
                        else if (!integer_fits(value, value_type) && value_type.kind != Type::Kind::Character)
                            fail(ErrorKind::InvalidInteger, "integer literal", "integer does not fit its declared scalar type");
                        else if (value_type.kind == Type::Kind::Character && (value.negative || value.magnitude.size() > 4U))
                            fail(ErrorKind::InvalidInteger, "character literal", "character payload exceeds unsigned 32-bit range");
                        return value;
                    }
                    case Type::Kind::Float16:
                    case Type::Kind::Float32:
                    case Type::Kind::Float64:
                    case Type::Kind::Float128:
                    {
                        const auto size = count(limits_.maximum_numeric_bytes, "floating literal length");
                        std::string spelling;
                        spelling.reserve(size);
                        for (std::size_t index = 0; index < size && !error_; ++index)
                        {
                            const auto value = byte("floating literal");
                            if (value > 0x7fU)
                                fail(ErrorKind::InvalidInteger, "floating literal", "floating spelling must be ASCII");
                            else
                                spelling.push_back(static_cast<char>(value));
                        }
                        if (!error_ && !floating_spelling_is_valid(spelling))
                            fail(ErrorKind::InvalidInteger, "floating literal", "floating spelling is not canonical");
                        return FloatingLiteral{ std::move(spelling) };
                    }
                    case Type::Kind::String:
                        return text("string literal");
                    case Type::Kind::Function:
                    case Type::Kind::Named:
                    case Type::Kind::TypeVariable:
                        fail(ErrorKind::UnsupportedType, "literal", "non-primitive literal type is invalid");
                        return std::monostate{};
                }
                return std::monostate{};
            }

            [[nodiscard]] auto
            atom() -> Atom
            {
                const auto tag = byte("atom tag");
                auto value_type = type();
                if (tag == 0U)
                    return Atom::variable(symbol("variable symbol"), std::move(value_type));
                if (tag == 1U)
                {
                    auto value = literal(value_type);
                    return Atom::constant(std::move(value), std::move(value_type));
                }
                fail(ErrorKind::InvalidTag, "atom tag", "unknown CorePrep atom tag");
                return {};
            }

            [[nodiscard]] auto
            operation_tag() -> Operation
            {
                const auto tag = byte("operation tag");
                if (tag > static_cast<std::uint8_t>(Operation::MakeClosure))
                {
                    fail(ErrorKind::InvalidTag, "operation tag", "unknown CorePrep operation tag");
                    return Operation::Copy;
                }
                return static_cast<Operation>(tag);
            }

            [[nodiscard]] auto
            capture_mode() -> CaptureMode
            {
                const auto tag = byte("capture mode");
                if (tag > static_cast<std::uint8_t>(CaptureMode::Unowned))
                {
                    fail(ErrorKind::InvalidTag, "capture mode", "unknown CorePrep capture mode");
                    return CaptureMode::Strong;
                }
                return static_cast<CaptureMode>(tag);
            }

            [[nodiscard]] auto
            capture() -> Capture
            {
                Capture value;
                value.mode = capture_mode();
                value.symbol = symbol("capture symbol");
                value.type = type();
                value.value = atom();
                return value;
            }

            void
            operation(Instruction &value)
            {
                value.operation = operation_tag();
                if (value.operation == Operation::MakeClosure)
                {
                    value.closure_function = symbol("closure function symbol");
                    value.captures = vector<Capture>(
                        limits_.maximum_operands_per_instruction,
                        "closure capture count",
                        [this] {
                            return capture();
                        });
                    return;
                }
                value.operands = vector<Atom>(limits_.maximum_operands_per_instruction, "operand count", [this] {
                    return atom();
                });
                if (error_)
                    return;
                if (value.operation == Operation::Copy && value.operands.size() != 1U)
                    fail(ErrorKind::InvalidCount, "copy", "copy must contain exactly one operand");
                if (value.operation == Operation::Call && value.operands.empty())
                    fail(ErrorKind::InvalidCount, "call", "call must contain a callee operand");
            }

            [[nodiscard]] auto
            instruction() -> Instruction
            {
                Instruction value;
                const auto tag = byte("instruction tag");
                if (tag > static_cast<std::uint8_t>(Instruction::Kind::Evaluate))
                {
                    fail(ErrorKind::InvalidTag, "instruction tag", "unknown CorePrep instruction tag");
                    return value;
                }
                value.kind = static_cast<Instruction::Kind>(tag);
                switch (value.kind)
                {
                    case Instruction::Kind::Bind:
                        value.destination = symbol("binding symbol");
                        value.type = type();
                        value.mutable_binding = boolean("binding mutability");
                        operation(value);
                        break;
                    case Instruction::Kind::Assign:
                        value.destination = symbol("assignment symbol");
                        value.operands.push_back(atom());
                        value.operation = Operation::Copy;
                        value.type = value.operands.empty() ? Type::unit() : value.operands.front().type;
                        break;
                    case Instruction::Kind::Evaluate:
                        operation(value);
                        break;
                }
                return value;
            }

            [[nodiscard]] auto
            terminator() -> Terminator
            {
                Terminator value;
                const auto tag = byte("terminator tag");
                if (tag > static_cast<std::uint8_t>(Terminator::Kind::Unreachable))
                {
                    fail(ErrorKind::InvalidTag, "terminator tag", "unknown CorePrep terminator tag");
                    return value;
                }
                value.kind = static_cast<Terminator::Kind>(tag);
                switch (value.kind)
                {
                    case Terminator::Kind::Return:
                        value.value = atom();
                        break;
                    case Terminator::Kind::Branch:
                        value.value = atom();
                        value.true_target = unsigned_integer<BlockId>("true target");
                        value.false_target = unsigned_integer<BlockId>("false target");
                        break;
                    case Terminator::Kind::Jump:
                        value.true_target = unsigned_integer<BlockId>("jump target");
                        break;
                    case Terminator::Kind::Unreachable:
                        break;
                }
                return value;
            }

            [[nodiscard]] auto
            block() -> Block
            {
                Block value;
                value.id = unsigned_integer<BlockId>("block id");
                value.instructions = vector<Instruction>(limits_.maximum_instructions_per_block, "instruction count", [this] {
                    return instruction();
                });
                value.terminator = terminator();
                return value;
            }

            [[nodiscard]] auto
            parameter() -> Parameter
            {
                return Parameter{ symbol("parameter symbol"), type() };
            }

            [[nodiscard]] auto
            function() -> Function
            {
                Function value;
                value.symbol = symbol("function symbol");
                value.parameters = vector<Parameter>(limits_.maximum_parameters_per_function, "parameter count", [this] {
                    return parameter();
                });
                value.return_type = type();
                value.entry = unsigned_integer<BlockId>("entry block");
                value.blocks = vector<Block>(limits_.maximum_blocks_per_function, "block count", [this] {
                    return block();
                });
                return value;
            }
        };
    } // namespace

    auto
    decode(std::span<const std::uint8_t> bytes, const Limits &limits) -> DecodeResult
    {
        return Reader(bytes, limits).document();
    }
} // namespace visual_xsharp::core::wire

File: Compiler/Haskell/Core/src/Visual/XSharp/Core/CorePrep/Wire/Decode.hs
-- SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
-- SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0
module Visual.XSharp.Core.CorePrep.Wire.Decode (decodeCorePrep, decodeCorePrepWith) where

import Data.Bits (Bits, shiftL, (.|.))
import Data.Char (chr)
import Data.Word (Word16, Word32, Word64, Word8)
import Visual.XSharp.AST
import Visual.XSharp.Core
import Visual.XSharp.Core.CorePrep
import Visual.XSharp.Core.CorePrep.Wire.Format

data DecoderState = DecoderState
    { remainingBytes :: [Word8]
    , consumedBytes :: Int
    , decoderLimits :: WireLimits
    }

newtype Decoder a = Decoder {runDecoder :: DecoderState -> Either WireError (a, DecoderState)}

instance Functor Decoder where
    fmap transform parser = Decoder $ \state -> do
        (value, next) <- runDecoder parser state
        pure (transform value, next)

instance Applicative Decoder where
    pure value = Decoder $ \state -> Right (value, state)
    functionParser <*> valueParser = Decoder $ \state -> do
        (function, afterFunction) <- runDecoder functionParser state
        (value, afterValue) <- runDecoder valueParser afterFunction
        pure (function value, afterValue)

instance Monad Decoder where
    parser >>= nextParser = Decoder $ \state -> do
        (value, afterValue) <- runDecoder parser state
        runDecoder (nextParser value) afterValue

decodeCorePrep :: [Word8] -> Either WireError CorePrepModule
decodeCorePrep = decodeCorePrepWith defaultWireLimits

decodeCorePrepWith :: WireLimits -> [Word8] -> Either WireError CorePrepModule
decodeCorePrepWith limits bytes
    | length bytes > maximumWireBytes limits =
        Left (wireError LimitExceeded 0 "wire byte length" "input exceeds configured byte limit")
    | otherwise = do
        (moduleValue, finalState) <- runDecoder decodeDocument (DecoderState bytes 0 limits)
        if null (remainingBytes finalState)
            then Right moduleValue
            else Left (wireError TrailingInput (consumedBytes finalState) "document" "bytes remain after CorePrep module")

decodeDocument :: Decoder CorePrepModule
decodeDocument = do
    magic <- takeBytes "magic" 4
    require (magic == wireMagic) InvalidMagic "magic" "input is not a Visual X# CorePrep wire document"
    version <- readWord16 "version"
    require
        (version == wireVersionNumber currentWireVersion)
        UnsupportedVersion
        "version"
        ("unsupported CorePrep wire version " ++ show version)
    flags <- readWord16 "flags"
    require (flags == 0) InvalidTag "flags" "reserved wire flags must be zero"
    decodeModule

decodeModule :: Decoder CorePrepModule
decodeModule = CorePrepModule <$> decodeQualifiedName <*> decodeVector "function count" maximumFunctions decodeFunction

decodeFunction :: Decoder CorePrepFunction
decodeFunction =
    CorePrepFunction
        <$> decodeResolvedName "function symbol"
        <*> decodeVector "parameter count" maximumParametersPerFunction decodeParameter
        <*> decodeType
        <*> decodeBlockId "entry block"
        <*> decodeVector "block count" maximumBlocksPerFunction decodeBlock

decodeParameter :: Decoder (ResolvedName, Type)
decodeParameter = (,) <$> decodeResolvedName "parameter symbol" <*> decodeType

decodeBlock :: Decoder CorePrepBlock
decodeBlock =
    CorePrepBlock
        <$> decodeBlockId "block id"
        <*> decodeVector "instruction count" maximumInstructionsPerBlock decodeInstruction
        <*> decodeTerminator

decodeInstruction :: Decoder CorePrepInstruction
decodeInstruction = do
    tag <- readWord8 "instruction tag"
    case tag of
        0 ->
            CorePrepBind
                <$> decodeResolvedName "binding symbol"
                <*> decodeType
                <*> decodeBool "binding mutability"
                <*> decodeOperation
        1 -> CorePrepAssign <$> decodeResolvedName "assignment symbol" <*> decodeAtom
        2 -> CorePrepEvaluate <$> decodeOperation
        _ -> invalidTag "instruction tag" tag

decodeOperation :: Decoder CorePrepOperation
decodeOperation = do
    tag <- readWord8 "operation tag"
    case tag of
        18 ->
            CorePrepMakeClosure
                <$> decodeResolvedName "closure function symbol"
                <*> decodeVector "closure capture count" maximumOperandsPerInstruction decodeCapture
        _ -> do
            atoms <- decodeVector "operand count" maximumOperandsPerInstruction decodeAtom
            case tag of
                0 -> case atoms of [atom] -> pure (CorePrepCopy atom); _ -> invalidArity "copy" 1 (length atoms)
                1 -> case atoms of callee : arguments -> pure (CorePrepCall callee arguments); [] -> invalidArity "call" 1 0
                _ -> case tagPrimitive tag of
                    Just primitive -> pure (CorePrepPrimitive primitive atoms)
                    Nothing -> invalidTag "operation tag" tag

decodeCapture :: Decoder CorePrepCapture
decodeCapture = do
    mode <- readWord8 "capture mode" >>= decodeCaptureMode
    name <- decodeResolvedName "capture symbol"
    valueType <- decodeType
    atom <- decodeAtom
    pure (CorePrepCapture mode name valueType atom)

decodeCaptureMode :: Word8 -> Decoder CaptureMode
decodeCaptureMode tag = case tag of
    0 -> pure StrongCapture
    1 -> pure WeakCapture
    2 -> pure UnownedCapture
    _ -> invalidTag "capture mode" tag

decodeAtom :: Decoder CorePrepAtom
decodeAtom = do
    tag <- readWord8 "atom tag"
    valueType <- decodeType
    case tag of
        0 -> CorePrepVariable <$> decodeResolvedName "variable symbol" <*> pure valueType
        1 -> CorePrepLiteral <$> decodeLiteral valueType <*> pure valueType
        _ -> invalidTag "atom tag" tag

decodeLiteral :: Type -> Decoder CoreLiteral
decodeLiteral valueType
    | valueType == unitType = pure CoreUnit
    | valueType == boolType = CoreBoolean <$> decodeBool "boolean literal"
    | valueType == stringType = CoreString <$> decodeText "string literal"
    | typeName valueType == "char" = CoreInteger <$> decodeInteger
    | typeName valueType `elem` integerTypeNames = CoreInteger <$> decodeInteger
    | typeName valueType `elem` floatingTypeNames = CoreFloating <$> decodeAscii "floating literal"
    | otherwise = failAt UnsupportedType "literal" "literal uses an unsupported type"

decodeTerminator :: Decoder CorePrepTerminator
decodeTerminator = do
    tag <- readWord8 "terminator tag"
    case tag of
        0 -> CorePrepReturn <$> decodeAtom
        1 -> CorePrepBranch <$> decodeAtom <*> decodeBlockId "true target" <*> decodeBlockId "false target"
        2 -> CorePrepJump <$> decodeBlockId "jump target"
        3 -> pure CorePrepUnreachable
        _ -> invalidTag "terminator tag" tag

decodeResolvedName :: String -> Decoder ResolvedName
decodeResolvedName context = do
    raw <- readWord64 context
    require (raw > 0) InvalidSymbol context "symbol id must be positive"
    require (raw <= fromIntegral (maxBound :: Int)) InvalidSymbol context "symbol id exceeds host Int range"
    spelling <- decodeText (context ++ " spelling")
    pure (ResolvedName (SymbolId (fromIntegral raw)) (Identifier spelling))

decodeQualifiedName :: Decoder QualifiedName
decodeQualifiedName =
    QualifiedName
        <$> decodeVectorWithMaximum
            "qualified name part count"
            65535
            (Identifier <$> decodeText "qualified name part")

decodeType :: Decoder Type
decodeType = decodeTypeAt 0

decodeTypeAt :: Int -> Decoder Type
decodeTypeAt depth = do
    limits <- currentLimits
    require (depth <= maximumTypeDepth limits) LimitExceeded "type" "type nesting exceeds configured limit"
    tag <- readWord8 "type tag"
    case tag of
        0 -> pure unitType
        1 -> pure boolType
        2 -> pure intType
        4 -> pure stringType
        3 -> pure (namedScalar "long")
        5 ->
            FunctionType
                <$> decodeVectorWithMaximum "function type parameter count" 65535 (decodeTypeAt (depth + 1))
                <*> decodeTypeAt (depth + 1)
        6 ->
            NamedType
                <$> decodeQualifiedName
                <*> decodeVectorWithMaximum "type argument count" 65535 (decodeTypeAt (depth + 1))
        7 -> TypeVariable <$> decodeResolvedName "type variable symbol"
        8 -> pure (namedScalar "char")
        9 -> pure (namedScalar "byte")
        10 -> pure (namedScalar "short")
        11 -> pure (namedScalar "longint")
        12 -> pure (namedScalar "ubyte")
        13 -> pure (namedScalar "ushort")
        14 -> pure (namedScalar "ulong")
        15 -> pure (namedScalar "uint")
        16 -> pure (namedScalar "ulongint")
        17 -> pure (namedScalar "sfloat")
        18 -> pure (namedScalar "lfloat")
        19 -> pure (namedScalar "float")
        20 -> pure (namedScalar "double")
        _ -> invalidTag "type tag" tag

namedScalar :: String -> Type
namedScalar name = NamedType (QualifiedName [Identifier name]) []

typeName :: Type -> String
typeName (NamedType (QualifiedName [Identifier name]) []) = name
typeName _ = ""

integerTypeNames :: [String]
integerTypeNames = ["byte", "short", "long", "int", "longint", "ubyte", "ushort", "ulong", "uint", "ulongint"]

floatingTypeNames :: [String]
floatingTypeNames = ["sfloat", "lfloat", "float", "double"]

decodeBlockId :: String -> Decoder Int
decodeBlockId context = fromIntegral <$> readWord32 context

decodeVector :: String -> (WireLimits -> Int) -> Decoder a -> Decoder [a]
decodeVector context selectLimit parser = do
    limits <- currentLimits
    decodeVectorWithMaximum context (selectLimit limits) parser

decodeVectorWithMaximum :: String -> Int -> Decoder a -> Decoder [a]
decodeVectorWithMaximum context maximumCount parser = do
    rawCount <- readWord32 context
    require
        (toInteger rawCount <= toInteger maximumCount)
        LimitExceeded
        context
        ("count " ++ show rawCount ++ " exceeds limit " ++ show maximumCount)
    sequence (replicate (fromIntegral rawCount) parser)

decodeText :: String -> Decoder String
decodeText context = do
    limits <- currentLimits
    count <- readWord32 (context ++ " length")
    require
        (toInteger count <= toInteger (maximumStringCodePoints limits))
        LimitExceeded
        context
        "text exceeds configured code point limit"
    sequence (replicate (fromIntegral count) (decodeCodePoint context))

decodeInteger :: Decoder Integer
decodeInteger = do
    negative <- decodeBool "integer sign"
    limits <- currentLimits
    count <- readWord32 "integer magnitude length"
    require
        (toInteger count <= toInteger (maximumNumericBytes limits))
        LimitExceeded
        "integer magnitude"
        "integer magnitude exceeds configured byte limit"
    magnitude <- takeBytes "integer magnitude" (fromIntegral count)
    require (canonicalMagnitude magnitude) InvalidInteger "integer magnitude" "integer magnitude is not canonical"
    require (not negative || not (null magnitude)) InvalidInteger "integer sign" "zero cannot have a negative sign"
    let value = foldl (\result octet -> result * 256 + fromIntegral octet) 0 magnitude
    pure (if negative then negate value else value)
    where
        canonicalMagnitude [] = True
        canonicalMagnitude (first : _) = first /= 0

decodeAscii :: String -> Decoder String
decodeAscii context = do
    limits <- currentLimits
    count <- readWord32 (context ++ " length")
    require
        (toInteger count <= toInteger (maximumNumericBytes limits))
        LimitExceeded
        context
        "numeric spelling exceeds configured byte limit"
    bytes <- takeBytes context (fromIntegral count)
    require (all (<= 0x7f) bytes) InvalidInteger context "numeric spelling must contain ASCII characters only"
    pure (map (chr . fromIntegral) bytes)

decodeCodePoint :: String -> Decoder Char
decodeCodePoint context = do
    value <- readWord32 context
    require
        (value <= 0x10ffff && not (value >= 0xd800 && value <= 0xdfff))
        InvalidCodePoint
        context
        "wire text contains a non-scalar Unicode code point"
    pure (chr (fromIntegral value))

decodeBool :: String -> Decoder Bool
decodeBool context = do
    value <- readWord8 context
    case value of
        0 -> pure False
        1 -> pure True
        _ -> failAt InvalidBoolean context ("invalid boolean byte " ++ show value)

tagPrimitive :: Word8 -> Maybe CorePrimitive
tagPrimitive tag = case tag of
    2 -> Just CoreAdd
    3 -> Just CoreSubtract
    4 -> Just CoreMultiply
    5 -> Just CoreDivide
    6 -> Just CoreFloorDivide
    7 -> Just CoreRemainder
    8 -> Just CoreLessThan
    9 -> Just CoreLessEqual
    10 -> Just CoreGreaterThan
    11 -> Just CoreGreaterEqual
    12 -> Just CoreEqual
    13 -> Just CoreNotEqual
    14 -> Just CoreLogicalAnd
    15 -> Just CoreLogicalOr
    16 -> Just CoreNegate
    17 -> Just CoreLogicalNot
    _ -> Nothing

invalidTag :: String -> Word8 -> Decoder a
invalidTag context tag = failAt InvalidTag context ("unknown tag " ++ show tag)

invalidArity :: String -> Int -> Int -> Decoder a
invalidArity context expected actual =
    failAt
        InvalidCount
        context
        ("expected " ++ show expected ++ " operand(s), found " ++ show actual)

currentLimits :: Decoder WireLimits
currentLimits = Decoder $ \state -> Right (decoderLimits state, state)

require :: Bool -> WireErrorKind -> String -> String -> Decoder ()
require condition kind context message = if condition then pure () else failAt kind context message

failAt :: WireErrorKind -> String -> String -> Decoder a
failAt kind context message = Decoder $ \state -> Left (wireError kind (consumedBytes state) context message)

takeBytes :: String -> Int -> Decoder [Word8]
takeBytes context count = Decoder $ \state ->
    let (prefix, suffix) = splitAt count (remainingBytes state)
     in if length prefix == count
            then Right (prefix, state {remainingBytes = suffix, consumedBytes = consumedBytes state + count})
            else
                Left
                    ( wireError
                        TruncatedInput
                        (consumedBytes state)
                        context
                        ("needed " ++ show count ++ " byte(s), found " ++ show (length prefix))
                    )

readWord8 :: String -> Decoder Word8
readWord8 context = do
    bytes <- takeBytes context 1
    case bytes of
        [value] -> pure value
        _ -> failAt TruncatedInput context "byte reader received an empty result"

readWord16 :: String -> Decoder Word16
readWord16 context = assemble <$> takeBytes context 2

readWord32 :: String -> Decoder Word32
readWord32 context = assemble <$> takeBytes context 4

readWord64 :: String -> Decoder Word64
readWord64 context = assemble <$> takeBytes context 8

assemble :: (Integral a, Bits a) => [Word8] -> a
assemble bytes = foldr combine 0 (zip [0 ..] bytes)
    where
        combine (index, value) result = result .|. (fromIntegral value `shiftL` (index * 8))
